core-bundle/src/Resources/contao/classes/FrontendTemplate.php

```diff
@@ -348,29 +348,38 @@ private function setCacheHeaders(Response $response)
 		/** @var $objPage PageModel */
 		global $objPage;
 
+		// Do not cache response if caching was not configured at all
 		if (($objPage->cache === false || $objPage->cache < 1) && ($objPage->clientCache === false || $objPage->clientCache < 1))
 		{
-			$response->headers->addCacheControlDirective('no-cache');
-			$response->headers->addCacheControlDirective('no-store');
+			return $this->makeResponsePrivate($response);
+		}
 
-			return $response->setPrivate();
+		// Do not cache response if: front end preview is active OR page is protected OR back end user is logged in
+		if (BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		{
+			return $this->makeResponsePrivate($response);
 		}
 
-		// Do not cache the response if a user is logged in or the page is protected
-		// TODO: Add support for proxies so they can vary on member context
-		if (FE_USER_LOGGED_IN === true || BE_USER_LOGGED_IN === true || $objPage->protected || $this->hasAuthenticatedBackendUser())
+		// Do not cache response if front end user is logged in UNLESS explicitly set in the page settings
+		if (FE_USER_LOGGED_IN === true && !$objPage->forceCache)
```

Comment:
IMO this should just be

```suggestion
		if (0 !== \count($response->headers->getCookies()))
```

So that we don’t store anything in the cache if it was triggered by a cookied request.